summary metrics parsing

timstclair opened the pull request:
This is still a work in progress, but "should" be fully functional. Feel free to comment on what I have so far.

For: https://github.com/kubernetes/heapster/issues/907

/cc @vishh @pwittrock

mwielgus commented:
As requested in #907, please leave the default kublet source as is. We want to reduce the risk of introducing bugs/breaking heapster just before the launch during addition of the new api support.

mwielgus commented on metrics/sources/kubelet/summary_decoders.go:
```diff
@@ -0,0 +1,238 @@
+		}
+
+		mv.ValueType = ValueFloat
+		mv.FloatValue = float32(metric.Value)
```
Are all the values float?

timstclair replied:
Yes

mwielgus replied:
I was under impression that cAdvisor's custom metrics were either int64 or float.

mwielgus commented on metrics/sources/kubelet/summary_decoders.go:
```diff
@@ -0,0 +1,238 @@
+	addIntMetric(metrics, &MetricCpuUsage, cpu.UsageCoreNanoSeconds, startTime)
+
+	// Need to translate from nanocores to millicores.
+	if cpu.UsageNanoCores != nil {
```
Is it guaranteed that we will always have cpu usage rate? How it is calculated?

timstclair replied:
Yes, it is straight from cAdvisor. Currently, it's calculated when the v1 API is translated into the v2 API ([here](https://github.com/google/cadvisor/blob/master/info/v2/conversion.go#L185)), but that will likely change in the future.

mwielgus commented:
cc: @piosz @fgrzadkowski

tallclair committed:
Add summary metrics source

Add an alternative to the `kubernetes` source: `kubernetes.summary`,
which uses the new /stats/summary endpoint when available (falling back
to the old source behavior with old (pre v1.2) kubelets).

timstclair commented:
I refactored this to minimize the changes to existing code, and make the Summary a separate metric source (as a wrapper around the kubeletMetricSource).

jimmidyson commented on metrics/sources/kubelet/summary.go:
```diff
@@ -0,0 +1,331 @@
+		}
+
+		mv.ValueType = ValueFloat
+		mv.FloatValue = float32(metric.Value)
```
Really a `float32`? Not limiting?

timstclair replied:
Ah, good point. I added some code to convert to an int64 when possible, but if it isn't I don't think we can do anything with the current API.

mwielgus commented:
Please make a full copy of kubelet source (name it kubelet2) and make your changes there. I know that the code will be duplicated but only for some time and I accept it. Once we are done with the new api we will delete the old source (kubelet) and rename kubelet2.

timstclair commented:
@mwielgus I think what I have here is almost the same thing (it doesn't touch `kubelet.go`), but without all the code duplication. My concern is that if a bug is found in the existing kubelet implementation, it would then need to be fixed in 2 places with your proposal. What is your issue with this implementation?

vishh commented:
@timstclair: Similar to network stats, can we add volume disk usage as well? We might need an additional field/label to qualify volume name.

timstclair commented:
@vishh - I forgot when we discussed this in the meeting, but volume stats are already added [here](https://github.com/kubernetes/heapster/pull/927/files#diff-e2b7cf1688b5dae5e9f990fc75ab1940R170). FS stats on a pod-scoped MetricSet can be interpreted as volume stats, but if you think it's necessary we could add a different label or prefix the fsKey (e.g. `vol:...`). I'm open to suggestions here.

vishh commented on metrics/sources/kubelet/summary.go:
```diff
@@ -0,0 +1,342 @@
+		metrics.LabeledMetrics = append(metrics.LabeledMetrics, usage)
+	}
+
+	if fs.CapacityBytes != nil {
```
What about `available`? Will these metrics be serialized and pushed to sinks automatically?

timstclair replied:
It looks like we need to add a new metric for available. I'm not sure I understand your second question?

vishh replied:
By sinks I meant the storage backends. Will these metrics be stored into
InfluxDB automatically or does it need a schema change somewhere?
One more want is to expose these new metrics via the GKE specific REST
endpoint. It can be done in a separate PR though to make it easy to review.

On Tue, Feb 2, 2016 at 3:41 PM Tim St. Clair <notifications@github.com>
wrote:

> In metrics/sources/kubelet/summary.go
> <https://github.com/kubernetes/heapster/pull/927#discussion_r51656180>:
>
> > +		return
> > +	}
> > +
> > +	fsLabels := map[string]string{LabelResourceID.Key: fsKey}
> > +	if fs.UsedBytes != nil {
> > +		usage := LabeledMetric{
> > +			Name:       MetricFilesystemUsage.Name,
> > +			Labels:     fsLabels,
> > +			ValueType:  ValueInt64,
> > +			MetricType: MetricFilesystemUsage.Type,
> > +			IntValue:   int64(*fs.UsedBytes),
> > +		}
> > +		metrics.LabeledMetrics = append(metrics.LabeledMetrics, usage)
> > +	}
> > +
> > +	if fs.CapacityBytes != nil {
>
> It looks like we need to add a new metric for available. I'm not sure I
> understand your second question?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/heapster/pull/927/files#r51656180>.
>

mwielgus replied:
Most of the sinks have some metric registration code in place if required. So in theory you can add any metric you like. Influxdb has a support for both normal and labeled metrics. GKE/GCM not yet but it is coming soon (probably tomorrow).

vishh replied:
Also should we update the metrics list [here](https://github.com/kubernetes/heapster/blob/heapster-scalability/metrics/core/metrics.go#L24) to include `available`?

vishh replied:
Are we planning to maintain a schema of some sort for all the metrics supported similar to [this](https://github.com/kubernetes/heapster/blob/master/docs/storage-schema.md)

timstclair replied:
Done. Re: schema - It looks like that schema is out of date (it's missing a handful of metrics). I can update it in a separate PR, but I think it would ideally be generated from the metrics definitions.

vishh replied:
Acknowledged.

On Wed, Feb 3, 2016 at 4:57 PM, Tim St. Clair <notifications@github.com>
wrote:

> In metrics/sources/kubelet/summary.go
> <https://github.com/kubernetes/heapster/pull/927#discussion_r51814713>:
>
> > +		return
> > +	}
> > +
> > +	fsLabels := map[string]string{LabelResourceID.Key: fsKey}
> > +	if fs.UsedBytes != nil {
> > +		usage := LabeledMetric{
> > +			Name:       MetricFilesystemUsage.Name,
> > +			Labels:     fsLabels,
> > +			ValueType:  ValueInt64,
> > +			MetricType: MetricFilesystemUsage.Type,
> > +			IntValue:   int64(*fs.UsedBytes),
> > +		}
> > +		metrics.LabeledMetrics = append(metrics.LabeledMetrics, usage)
> > +	}
> > +
> > +	if fs.CapacityBytes != nil {
>
> Done. Re: schema - It looks like that schema is out of date (it's missing
> a handful of metrics). I can update it in a separate PR, but I think it
> would ideally be generated from the metrics definitions.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/heapster/pull/927/files#r51814713>.
>

mwielgus commented on metrics/sources/factory.go:
```diff
@@ -30,6 +30,12 @@ func (this *SourceFactory) Build(uri flags.Uri) (core.MetricsSourceProvider, err
 	case "kubernetes":
 		provider, err := kubelet.NewKubeletProvider(&uri.Val)
 		return provider, err
+	case "summary":
```
Name the sink so that it has something in common with kubernetes/kubelet/etc.

timstclair replied:
Done. Is the `.` delimiter acceptable?

mwielgus commented on metrics/sources/kubelet/summary.go:
```diff
@@ -0,0 +1,342 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
```
Place this file in a separate directory. One source per dir, please.

timstclair replied:
Responded here: https://github.com/kubernetes/heapster/pull/927#issuecomment-179498431

mwielgus commented:
@timstclair Ok, if these are the only changes to the old source. But move summary.go to a separate dir/package.

timstclair commented:
> Ok, if these are the only changes to the old source. But move summary.go to a separate dir/package.

Unfortunately it's a tradeoff. The current wrapper approach reads some private fields from the Kubelet source. To move the summary source to a new package will either require more duplicate code or more changes to existing code (export internal fields). Personally I like the current approach for the balance it makes between (1) Isolating the new source, (2) minimal changes to existing code and (3) minimal duplicate code.

Thoughts?

*On a side note, I'm happy to own the cleanup of this post-1.2*

vishh commented on metrics/sources/kubelet/summary.go:
```diff
@@ -0,0 +1,342 @@
+		case stats.MetricCumulative:
+			mv.MetricType = MetricCumulative
+		default:
+			glog.V(4).Infof("Skipping %s: unknown custom metric type: %v", metric.Name, metric.Type)
```
Will rate or delta metrics be silently dropped?

timstclair replied:
As it currently stands, yes. I suppose I should just add a new `MetricType` ([here](https://github.com/kubernetes/heapster/blob/heapster-scalability/metrics/core/types.go)). Are there any consequences to simply adding a new type that I need to be aware of?

vishh replied:
Some of the storage backends use the metric type as a hint. Google Cloud Monitoring on the other hand does recognize `Delta` as an explicit type of metric

timstclair replied:
Done.

vishh commented on metrics/sources/kubelet/summary.go:
```diff
@@ -0,0 +1,342 @@
+
+		const epsilon = 1e-9 // Margin of error
+		_, frac := math.Modf(math.Abs(metric.Value))
+		if metric.Value < math.MaxInt64 && (frac < epsilon || frac > 1.0-epsilon) {
```
Given that the value is float64 wouldn't be always less than `MaxInt64`?

timstclair replied:
No. Max float64 is order 10^308, while max int64 is around 10^19.

vishh replied:
:+1:

vishh commented on metrics/sources/kubelet/summary.go:
```diff
@@ -0,0 +1,342 @@
+		if metric.Value < math.MaxInt64 && (frac < epsilon || frac > 1.0-epsilon) {
+			mv.ValueType = ValueInt64
+			mv.IntValue = int64(metric.Value)
+			if frac > 1.0-epsilon {
```
Should we instead round up if `frac` is greater than `0.5`?

timstclair replied:
It's logically equivalent here, but if you think using 0.5 is more readable, sure :)

vishh replied:
Isn't `1.0-epsilon` a number that is tending to `1`? Can you explain why it is logically equivalent?

timstclair replied:
Because the execution can only get here if `(frac < epsilon || frac > 1.0-epsilon)`, so if `frac > 0.5` then it follows that `frac > 1.0-epsilon`.

mwielgus replied:
I would rather suggest that you stick to one type of the value for a metric. We can consider moving from float32 to float64 in MV using extra bytes saved from switching Time.Time (which is 20b) to int32 with unix timestamp.

timstclair replied:
Ok, sgtm. Done.

vishh commented on metrics/sources/kubelet/summary.go:
```diff
@@ -0,0 +1,342 @@
+			}
+		} else {
+			mv.ValueType = ValueFloat
+			mv.FloatValue = float32(metric.Value)
```
If float values are supported, why are we converting to an integer in the first place?

timstclair replied:
Because only float32 values are supported, which only have 23 bits of precision. The integer conversion is an attempt to offer higher precision for integer values (such as counts). Do you think this is overcomplicating the conversion?

vishh replied:
My bad. I overlooked the float32() cast. This logic LGTM

vishh replied:
BTW, GCM supports double values. Should we avoid limiting precision in heapster core and instead let the storage backends deal with the precision?
That way we can store doubles here.

timstclair replied:
I'm trying to minimize heapster changes in this PR. I think that's a good idea, but I'd prefer to handle it in a follow-up. I'll add a TODO.

vishh commented on metrics/sources/kubelet/summary.go:
```diff
@@ -0,0 +1,342 @@
+	this.decodeUptime(podMetrics, startTime)
+	this.decodeNetworkStats(podMetrics, pod.Network, startTime)
+	for _, vol := range pod.VolumeStats {
+		this.decodeFsStats(podMetrics, vol.Name, &vol.FsStats)
```
Should the key include the keyword `vol` in there?

timstclair replied:
Done.

vishh commented:
Completed my first pass through this PR.

timstclair commented:
Ok, addressed feedback & finished testing. I added a unit test, and an alternative controller to deploy the new source for the integration test (and verified it works). I think the main remaining issue is that of separating the summary source to a different directory, but I described the tradeoffs above.

PTAL - We'd really like to get this into the 1.2 release.

timstclair commented:
Also I'm aware that the Godeps changes make this hard to review. Let me know if you'd like me to separate them and/or squash the commits.

vishh commented:
@timstclair: What's the status on this PR?

timstclair commented:
Rebased & squashed non-godeps commits.

timstclair commented:
> @timstclair: What's the status on this PR?

From my perspective, this is good to go. I'm just waiting on a review.

timstclair commented:
Addressed some comments, responses on https://github.com/timstclair/heapster/commit/d78bdc30025871f8a21b3291ea4fd040f0bb6c45

Will ping again once I've addressed the last 2 (side-by-side testing & smooth CPU usage)

dchen1107 commented:
https://github.com/kubernetes/kubernetes/pull/20990 was merged. Anything else is blocking this one?

timstclair commented:
I just need to pull in the new godeps & update this PR to use the timestamps added in 
https://github.com/kubernetes/kubernetes/pull/20990

tallclair committed:
Update heapster k8s Godeps to include latest stats summary API

timstclair commented:
I can't find the open comment for some reason, but I've addressed the cpu rate issues in the latest push. A few things to note:
- This will need to be synced with https://github.com/kubernetes/heapster/pull/985, and the temporary code can be removed
- This includes a temporary Godeps update - the real update is blocked on https://github.com/kubernetes/kubernetes/pull/21368 (without it most of kubernetes & cAdvisor will need to be pulled in to heapster)

PTAL - hopefully we can work out all the remaining issues by EOD tomorrow so this will be in friday's build.

mwielgus commented on deploy/kube-config/standalone-test/heapster-summary-controller.yaml:
```diff
@@ -0,0 +1,44 @@
+apiVersion: v1
```
Which yaml is used in integration tests?

timstclair replied:
It still defaults to heapster.controller.yaml, but it's configurable with a [flag](https://github.com/kubernetes/heapster/blob/heapster-scalability/integration/heapster_api_test.go#L44)

mwielgus commented on metrics/sources/kubelet/summary.go:
```diff
@@ -0,0 +1,367 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubelet
```
Can we please have it in a separate package?

timstclair replied:
Done.

mwielgus commented on metrics/sources/kubelet/summary.go:
```diff
@@ -0,0 +1,367 @@
+			mv.MetricType = MetricGauge
+		case stats.MetricCumulative:
+			mv.MetricType = MetricCumulative
+		case stats.MetricDelta:
```
Add delta handling in pod aggragtor. We should not aggregate delta metrics as we don't aggregate cumulatives.

timstclair replied:
Done.

mwielgus commented on metrics/sources/kubelet/summary.go:
```diff
@@ -0,0 +1,367 @@
+
+func (this *summaryMetricsSource) decodeNodeStats(metrics map[string]*MetricSet, labels map[string]string, node *stats.NodeStats) {
+	nodeMetrics := &MetricSet{
+		Labels:         this.cloneLabels(labels),
```
set create and scrape time

timstclair replied:
Done.

timstclair commented:
Adressed comments.

mwielgus commented:
LGTM

mwielgus commented:
do not merge yet (per chat with @timstclair )

timstclair commented:
Updated with the real godeps & squashed. Unit & integration tests pass. Ready to merge!

mwielgus commented:
all green, merging

timstclair commented:
FYI, integration tests pass for all combinations of:
SUPPORTED_KUBE_VERSIONS={1.1.7,HEAD}
source={kubelet, summary_api}
